Accenture/reactive-interaction-gateway
# Use dedicated Kafka groups to allow multi-purpose Kafka topics

## Description

Using the same Kafka group everywhere means that only one of the consumers will
get the Kafka message. However, in RIG each Kafka handler represents a different
feature and they must not affect each other.

For KAFKA_TO_FILTER and KAFKA_TO_HTTP this is actually a bugfix, because both
handlers would typically be subscribed to the same topic, yet in that case only
one of them would receive its messages.

For PROXY_KAFKA_RESPONSE_TOPICS this change allows sharing the Kafka topic with
the other consumers as well. However, since this causes RIG to process each
event at least twice, this is not recommended.

Closes #206.

## What to look out for

Dear reviewer, I want you to

- check that the code works on your machine (instructions below).
- suggest documentation improvements. A user typically shouldn't need to alter the group IDs, so I'm not sure we should add that to the docs. What do you think?

## How to test this

By default, an event and an async response should be processed only once each:

1. start kafka
1. start rig with default config
1. send an invalid kafka message to the "rig" topic: `echo test | kafkacat -b localhost:9092 -t rig`
1. you should see an error coming from application=rig_kafka module=RigKafka.Client.GroupSubscriber
1. send an invalid kafka message to the "rig-proxy-response" topic: `echo test | kafkacat -b localhost:9092 -t rig-proxy-response`
1. you should see a warning coming from application=rig_inbound_gateway module=RigInboundGateway.ApiProxy.Handler.Kafka

Restart RIG with `ROXY_KAFKA_RESPONSE_TOPICS=rig`, then:

1. `echo test | kafkacat -b localhost:9092 -t rig`
1. => you should see both of the previous messages, which means both handlers process the message.
